sled-agent/src/sled_agent.rs

```diff
@@ -948,9 +948,48 @@ impl SledAgent {
         Ok(disk_result)
     }
 
+    /// Ensures that the specific sets of disks, datasets, and zones specified
+    /// by `config` are running.
+    ///
+    /// This method currently blocks while each of disks, datasets, and zones
+    /// are ensured in that order; a failure on one prevents any attempt to
+    /// ensure the subsequent step(s).
+    pub async fn set_omicron_config(
+        &self,
+        config: OmicronSledConfig,
+    ) -> Result<OmicronSledConfigResult, Error> {
+        let disks =
+            self.omicron_physical_disks_ensure(config.disks_config).await?;
+
+        // If we only had partial success deploying disks, don't proceed.
+        if disks.has_error() {
+            return Ok(OmicronSledConfigResult {
+                disks: disks.status,
+                datasets: Vec::new(),
+            });
+        }
+
+        let datasets = self.datasets_ensure(config.datasets_config).await?;
+
+        // If we only had partial success deploying datasets, don't proceed.
+        if datasets.has_error() {
+            return Ok(OmicronSledConfigResult {
+                disks: disks.status,
+                datasets: datasets.status,
+            });
+        }
+
+        self.omicron_zones_ensure(config.zones_config).await?;
```

Comment:
Why does "bad applied config" of disks/datasets return a `OmicronSledConfigResult`, but for zones, we return an error?

Put another way: Why doesn't `OmicronSledConfigResult` contain a "zone result" field? It's supposed to be able to give info about partial errors, which may include "we could deploy all disks and datasets, but failed to deploy zones", right?

(I acknowledge this might not matter as we prep to have this method be asynchronous, but it felt a little lopsided)

Reply:
Short answer: Because that's how the existing endpoints work (as well as the internal sled-agent methods, right? ensuring omicron zones doesn't return any kind of partial success result). I was trying to make no changes other than those required to mechanically join the endpoints together.